man/ja/generate.1
'\" t
.\"     Title: 
    generate
  
.\"    Author: 
.\" Generator: DocBook XSL Stylesheets v1.76.1 <http://docbook.sf.net/>
.\"      Date: 19/12/2013
.\"    Manual: 
.\"    Source: 
.\"  Language: Japanese
.\"
.TH "GENERATE" "1" "19/12/2013" ""
.\" -----------------------------------------------------------------
.\" * Define some portability stuff
.\" -----------------------------------------------------------------
.\" ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
.\" http://bugs.debian.org/507673
.\" http://lists.gnu.org/archive/html/groff/2009-02/msg00013.html
.\" ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
.ie \n(.g .ds Aq \(aq
.el       .ds Aq '
.\" -----------------------------------------------------------------
.\" * set default formatting
.\" -----------------------------------------------------------------
.\" disable hyphenation
.nh
.\" disable justification (adjust text to left margin only)
.ad l
.\" -----------------------------------------------------------------
.\" * MAIN CONTENT STARTS HERE *
.\" -----------------------------------------------------------------
.SH "名前"

    generate
   \- 文法から文をランダム生成する
.SH "概要"
.HP \w'\fBgenerate\fR\ 'u
\fBgenerate\fR [\-v] [\-t] [\-n\ \fInum\fR] [\-s\ \fIspname\fR] {prefix}
.SH "DESCRIPTION"
.PP
generate は文法に従って文をランダムに生成します．
.PP
実行には \&.dfa, \&.dict, \&.term の各ファイルが必要です． あらかじめ
\fBmkdfa\&.pl\fR
で生成しておいて下さい．
.SH "OPTIONS"
.PP
\fB \-t \fR
.RS 4
単語ではなくカテゴリ名で出力する．
.RE
.PP
\fB \-n \fR \fInum\fR
.RS 4
生成する文の数を指定する (default: 10)
.RE
.PP
\fB \-s \fR \fIspname\fR
.RS 4
生成においてスキップすべきショートポーズ単語の名前を指定する． (default: "sp")
.RE
.PP
\fB \-v \fR
.RS 4
デバッグ出力．
.RE
.SH "EXAMPLES"
.PP
vfr (フィッティングタスク用文法) での実行例：
.sp
.if n \{\
.RS 4
.\}
.nf
% generate vfr
Reading in dictionary\&.\&.\&.done
Reading in DFA grammar\&.\&.\&.done
Mapping dict item <\-> DFA terminal (category)\&.\&.\&.done
Reading in term file (optional)\&.\&.\&.done
42 categories, 99 words
DFA has 135 nodes and 198 arcs
 \-\-\-\-\- 
silB やめます silE
silB 終了します silE
silB シャツ を スーツ と 統一して 下さい silE
silB スーツ を カッター と 同じ 色 に 統一して 下さい silE
silB 交換して 下さい silE
silB これ を 覚えておいて 下さい silE
silB 覚えておいて 下さい silE
silB 戻って 下さい silE
silB スーツ を シャツ と 統一して 下さい silE
silB 上着 を 橙 に して 下さい silE
.fi
.if n \{\
.RE
.\}
.sp
.SH "SEE ALSO"
.PP

\fB mkdfa.pl \fR( 1 )
,
\fB generate-ngram \fR( 1 )
.SH "COPYRIGHT"
.PP
Copyright (c) 1991\-2013 京都大学 河原研究室
.PP
Copyright (c) 1997\-2000 情報処理振興事業協会(IPA)
.PP
Copyright (c) 2000\-2005 奈良先端科学技術大学院大学 鹿野研究室
.PP
Copyright (c) 2005\-2013 名古屋工業大学 Julius開発チーム
.SH "LICENSE"
.PP
Julius の使用許諾に準じます．


ChangeLog
2008-05-17 13:06  sumomo

	* libjulius/src/: instance.c, m_fusion.c, m_options.c: Now GMM
	  module uses the analysis parameter of the given AM when "-AM_GMM"
	  is not explicitly specified.	Now output error when GMM has
	  different sampling rate from AM.

2008-05-13 15:32  sumomo

	* jcontrol/jcontrol.c: Flush output after each message.

2008-05-13 15:32  sumomo

	* libsent/src/util/readfile.c: Fixed file name prompt to ignore
	  empty input correctly.

2008-05-09 15:00  sumomo

	* libjulius/src/m_usage.c: Grouped the usage messages by the
	  belonging modules.

2008-05-09 14:59  sumomo

	* libjulius/src/m_options.c: New option "-usepower" to use power
	  instead of magnitude in filterbank analysis.

2008-05-09 14:58  sumomo

	* libsent/: include/sent/mfcc.h, src/hmminfo/read_binhmm.c,
	  src/hmminfo/write_binhmm.c, src/wav2mfcc/mfcc-core.c: Added
	  support for USEPOWER in filterbank analysis.

2008-05-09 14:57  sumomo

	* libsent/src/wav2mfcc/para.c: Add support for USEPOWER.  Fixed
	  -htkconf to assume SOURCERATE=625 (16kHz) if not exist.

2008-05-08 14:45  sumomo

	* libjulius/: include/julius/extern.h, src/adin-cut.c,
	  src/m_adin.c: Add tiny value check for frequency.

2008-05-08 14:44  sumomo

	* libsent/src/adin/adin_mic_linux_alsa.c: Fixed failure to startup
	  when using multiple sound cards on ALSA.

2008-05-08 14:43  sumomo

	* libsent/src/adin/adin_file.c: Fixed segfault when failed to open
	  audio file.

2008-04-27 22:06  sumomo

	* libjulius/: include/julius/jconf.h, src/default.c, src/m_info.c,
	  src/m_options.c, src/m_usage.c, src/search_bestfirst_main.c: Add
	  an option "-fallback1pass" to use the 1st pass result as final
	  result when the 2nd pass fails.  This was the default on 3.x.

2008-04-27 22:06  sumomo

	* libjulius/src/version.c.in: Update copyright output.

2008-03-28 13:52  sumomo

	* libsent/src/adin/adin_mic_linux_alsa.c: Fix ALSA A/D-in to use
	  device other than "default".

2008-03-28 12:57  sumomo

	* julius/record.c: Fixed record sometimes fails.

2008-03-28 12:51  sumomo

	* libjulius/src/adin-cut.c: Moved CALLBACK_ADIN_TRIGGERED to adin
	  thread.

2008-03-25 17:12  sumomo

	* libjulius/src/pass1.c: Fixed GMM rejection not working well on
	  microphone input (enbuged on 4.0.1).

2008-03-21 11:59  sumomo

	* libsent/src/wav2mfcc/para.c: Add warning when HTK Config given by
	  "-htkconf" has unsupported values.

2008-03-20 18:08  sumomo

	* libsent/: include/sent/ngram2.h, include/sent/util.h,
	  src/ngram/ngram_access.c, src/ngram/ngram_compact_context.c,
	  src/ngram/ngram_read_arpa.c, src/ngram/ngram_read_bin.c,
	  src/ngram/ngram_util.c, src/ngram/ngram_write_bin.c,
	  src/util/endian.c: Update N-gram codes to (hopefully) allow 64bit
	  index in the future.

2008-03-20 18:06  sumomo

	* julius/: charconv_win32.c, libjcode/jlib.h, libjcode/libjcode.c:
	  Fixed euc-jp conversion not working on some string.

2008-03-18 02:00  sumomo

	* libjulius/src/: adin-cut.c, recogmain.c: Fixed
	  CALLBACK_EVENT_SPEECH_* call timing when -spsegment/VAD enabled.

2008-03-18 01:59  sumomo

	* libjulius/src/version.c.in, libsent/src/util/confout.c: Fix
	  typos.

2008-03-18 01:58  sumomo

	* julius/main.c, julius/output_stdout.c,
	  libjulius/include/julius/global.h, libjulius/src/callback.c,
	  libjulius/src/m_options.c, libjulius/src/m_usage.c:
	  "-callbackdebug" moved from Julius to JuliusLib.

2008-03-14 22:28  sumomo

	* libsent/src/ngram/ngram_compact_context.c: Fixed a bug that
	  prevents reading big N-gram.

2008-03-12 15:43  tag release_4_0_1

2008-03-12 15:43  sumomo

	* 00readme-ja.txt, 00readme.txt, ChangeLog, LICENSE.txt,
	  Release-ja.txt, Release.txt, libjulius/configure,
	  libjulius/configure.in, libsent/configure, libsent/configure.in,
	  support/build-all.sh: Release of 4.0.1.

2008-03-12 14:48  sumomo

	* libjulius/src/pass1.c: Fixed compilation failure enbuged at last
	  fix...

2008-03-10 11:39  sumomo

	* libjulius/src/: pass1.c, realtime-1stpass.c, recogmain.c: Fix CMN
	  not working well on spsegment/GMM-VAD/decoder-VAD mode.

2008-03-02 02:10  sumomo

	* libsent/: configure, configure.in, include/sent/config.h.in,
	  src/adin/adin_mic_linux_alsa.c: Fixed ALSA recording to work.
	  Many codes are updated.

2008-02-21 13:10  sumomo

	* libjulius/src/beam.c: Word recognition results are not correctly
	  sorted when using "-output N".

2008-02-18 23:35  sumomo

	* julius/module.c: Added a module command "GRAMINFO" to return
	  grammar info.

2008-02-18 23:32  sumomo

	* libjulius/src/m_fusion.c: Now -spsegment detects error when pause
	  word is not defined in dictionary.

2008-02-15 03:26  sumomo

	* libsent/src/phmm/mkwhmm.c: Fixed bug of immediate stop when no sp
	  model in multi-path mode.

2008-02-15 03:25  sumomo

	* libjulius/src/: m_info.c, search_bestfirst_v1.c,
	  search_bestfirst_v2.c, word_align.c: Fixed hungup when sp model
	  not exist on AM in multipath.

2008-02-14 13:44  sumomo

	* libjulius/: include/julius/extern.h, src/multi-gram.c: Changes
	  for better grammar handling: - multigram_add() returns the newly
	  assigned grammar id.	- a function is added to get grammar id
	  from name string.  - modified STAT message to output grammar name
	  together with id.

2008-02-14 13:41  sumomo

	* julius/module.c: Extended ACTIVATEGRAM, DEACTIVATEGRAM, DELGRAM
	  to specify grammar by name.

2008-02-13 17:49  sumomo

	* libjulius/include/julius/extern.h: Updated for some function
	  changes.

2008-02-13 17:48  sumomo

	* libjulius/src/: outprob_style.c, search_bestfirst_v1.c,
	  search_bestfirst_v2.c: Global variables and work area has been
	  moved to RecogProcess and StackDecode.

2008-02-13 17:47  sumomo

	* libjulius/src/wchmm.c: qsort function switched to qsort_reentrant
	  in libsent for modularity.

2008-02-13 17:46  sumomo

	* libjulius/src/confnet.c: Updated for being re-entrant.

2008-02-13 17:45  sumomo

	* libjulius/src/: m_jconf.c, m_options.c, ngram_decode.c,

gramtools/generate/generate.c
/*
 * Copyright (c) 1991-2013 Kawahara Lab., Kyoto University
 * Copyright (c) 2000-2005 Shikano Lab., Nara Institute of Science and Technology
 * Copyright (c) 2005-2013 Julius project team, Nagoya Institute of Technology
 * All rights reserved
 */

/* generate --- generate random sentences acceptable by given grammar */

#include "common.h"
#include "gen_next.h"
#if defined(_WIN32) && !defined(__CYGWIN32__)
#include "process.h"
#endif

#define MAXHYPO 300

WORD_INFO *winfo;
DFA_INFO *dfa;
char **termname;
boolean verbose_flag = FALSE;
boolean term_mode = FALSE;
boolean no_term_file;

NODE *
new_generate()
{
  NEXTWORD **nw;
  NODE *now;
  int i,j,num,selected;

  /* init */
  nw = nw_malloc();
  now = (NODE *)mymalloc(sizeof(NODE));
  now->endflag = FALSE;
  now->seqnum = 0;

  /* set init hypo */
  if (term_mode) {
    num = dfa_firstterms(nw);
  } else {
    num = dfa_firstwords(nw);
  }

  for (;;) {
    if (verbose_flag) {
      if (no_term_file) {
	for(i=0;i<num;i++)printf("\t-> %s\t%s\n",winfo->wname[nw[i]->id],winfo->woutput[nw[i]->id]);
      } else {
	for(i=0;i<num;i++)printf("\t-> %s\t%s\n",termname[winfo->wton[nw[i]->id]],winfo->woutput[nw[i]->id]);
      }
    }
    /* select random one */
    if (num == 1) {
      selected = 0;
    } else {
      j = abs(rand()) % num;
      for(i=0;i<j;i++) {
	selected = abs(rand()) % num;
      }
    }
    if (selected >= num) selected = num - 1;
    
    now->seq[now->seqnum++] = nw[selected]->id;
    now->state = nw[selected]->next_state;

    if (now->seqnum >= MAXSEQNUM) {
      printf("word num exceeded %d\n", MAXSEQNUM);
      nw_free(nw);
      return(now);
    }

    /* output */
    if (verbose_flag) {
      printf("(%3d) %s\n", now->state, winfo->woutput[now->seq[now->seqnum-1]]);
    }

    /* end check */
    if (dfa_acceptable(now)) break;

    /* get next words */
    if (term_mode) {
      num = dfa_nextterms(now, nw);
    } else {
      num = dfa_nextwords(now, nw);
    }
  }
  
  nw_free(nw);
  return(now);
  
}

static boolean
match_node(NODE *a, NODE *b)
{
  int i;
  
  if (a->seqnum != b->seqnum) return(FALSE);
  for (i=0;i<a->seqnum;i++) {
    if (a->seq[i] != b->seq[i]) return(FALSE);
  }
  return(TRUE);
}

static void
generate_main(int num)
{
  NODE *sent;
  NODE **stock;
  int i,n,c;

  /* avoid generating same sentence */
  stock = (NODE **)mymalloc(sizeof(NODE *)*num);
  n = 0;
  c = 0;
  while (n < num) {
    sent = new_generate();
    for (i=0;i<n;i++) {
      if (match_node(sent, stock[i])) break;
    }
    if (i >= n) {		/* no match, store as new */
      stock[n++] = sent;
      for (i=sent->seqnum-1;i>=0;i--) {
	if (term_mode) {
	  if (no_term_file) {
	    printf(" %s", winfo->wname[sent->seq[i]]);
	  } else {
	    printf(" %s", termname[winfo->wton[sent->seq[i]]]);
	  }
	} else {
	  printf(" %s", winfo->woutput[sent->seq[i]]);
	}
      }
      printf("\n");
      c = 0;
    } else {			/* same, ignored */
      c++;
      if (c >= MAXHYPO) {
	printf("no further sentence in the last %d trial\n", c);
	break;
      }
      free(sent);
    }
  }
  
  for(i=0;i<n;i++) free(stock[i]);
  free(stock);
}


static char *
usage(char *s)
{
  fprintf(stderr, "generate --- sentence random generator\n");
  fprintf(stderr, "usage: %s [-v] [-n] prefix\n",s);
  fprintf(stderr, "  -n num  ... generate N sentences (default: 10)\n");
  fprintf(stderr, "  -t      ... use category symbols instead of words (needs .term)\n");
  fprintf(stderr, "  -s string ... specify short-pause model\n");
  fprintf(stderr, "  -v      ... verbose output\n");
  exit(1);
}

static void
put_dfainfo()
{
  printf("%d categories, %d words\n",dfa->term_num,winfo->num);
  printf("DFA has %d nodes and %d arcs\n", dfa->state_num, dfa->arc_num);
}


int main(int argc, char *argv[])
{
  int i, len;
  char *prefix = NULL;
  char *dfafile, *dictfile, *termfile;
  int gnum = 10;
  char *spname_default = SPNAME_DEF;
  char *spname = NULL;
#define NEXTARG (++i >= argc) ? (char *)usage(argv[0]) : argv[i]

  /* argument */
  for(i=1;i<argc;i++) {
    if (argv[i][0] == '-') {
      switch(argv[i][1]) {
      case 'v':			/* verbose output */
	verbose_flag = TRUE;
	gnum = 1;
	break;
      case 't':			/* terminal mode */
	term_mode = TRUE;
	break;
      case 'n':
	gnum = atoi(NEXTARG);
	break;
      case 's':
	if (++i >= argc) {
	  usage(argv[0]);
	}
	spname = argv[i];
	break;
      default:
	fprintf(stderr, "no such option: %s\n",argv[i]);
	usage(argv[0]);
      }
    } else {
      prefix = argv[i];
    }
  }
  if (prefix == NULL) usage(argv[0]);

  if (spname == NULL) spname = spname_default;

  len = strlen(prefix) + 10;
  dfafile = (char *)mymalloc(len);
  dictfile = (char *)mymalloc(len);
  termfile = (char *)mymalloc(len);
  strcpy(dfafile, prefix);
  strcat(dfafile, ".dfa");
  strcpy(dictfile, prefix);
  strcat(dictfile, ".dict");
  strcpy(termfile, prefix);
  strcat(termfile, ".term");

  /* start init */
  winfo = word_info_new();
  init_voca(winfo, dictfile, NULL, TRUE, FALSE);
  dfa = dfa_info_new();
  init_dfa(dfa, dfafile);
  make_dfa_voca_ref(dfa, winfo);

  termname = (char **)mymalloc(sizeof(char *) * dfa->term_num);
  init_term(termfile, termname);
  if (termname[0] == NULL) {	/* no .term file */
    no_term_file = TRUE;
  } else {
    no_term_file = FALSE;
  }

  /* output info */
  put_dfainfo();

  /* set dfa->sp_id and dfa->is_sp[cid] from name "sp" */
  {
    int t, i;
    WORD_ID w;

    dfa->sp_id = WORD_INVALID;
    dfa->is_sp = (boolean *)mymalloc(sizeof(boolean) * dfa->term_num);
    for(t=0;t<dfa->term_num;t++) {
      dfa->is_sp[t] = FALSE;
      for(i=0;i<dfa->term.wnum[t]; i++) {
	w = dfa->term.tw[t][i];
	if (strcmp(winfo->woutput[w], spname) == 0) {
	  if (dfa->sp_id == WORD_INVALID) dfa->sp_id = w;
	  dfa->is_sp[t] = TRUE;
	  break;
	}
      }
    }
  }
  if (verbose_flag) {
    if (dfa->sp_id != WORD_INVALID) {
      printf("skippable word for NOISE: %s\t%s\n", winfo->wname[dfa->sp_id], winfo->woutput[dfa->sp_id]);
    }
  }
  printf("----- \n");

  /* random seed */
  srand(getpid());

  /* main loop */
  generate_main(gnum);

  free(dfafile);
  free(dictfile);
  return 0;
}
